### PR TITLE
Poste: remove support for comma-separated addresses

### PR DESCRIPTION
**Pull request by breville**
This removes support for splitting multiple addresses in a field by comma, since an audit of our entire `poste_deliveries` table shows that we never used the functionality.

We now log a Honeybadger error if we do detect a comma provided outside of surrounding double-quotes, helping us to catch any attempt at using that functionality.

### CODE REVIEW

**Comment by islemaster** [comment]
Does this change obsolete https://github.com/code-dot-org/code-dot-org/pull/29192 ?  Should we close that PR?

File: bin/cron/deliver_poste_messages [hunk 1]
```diff
@@ -245,20 +245,37 @@ class Deliverer
...
-      else
-        {name: nil, email: user_string}
-      end
+  def parse_address_string(string)
+    # A comma without surrounding double-quotes isn't necessarily bad, since it
+    # might be like this:
+    #   friendly name, phd <email@code.org>
+    #
+    # Though in such a case, it should really be provided like this:
+    #   "friendly name, phd", <email@code.org>
```
Comment (islemaster) [inline bin/cron/deliver_poste_messages]: Just making sure, was the additional comma inserted in this example intentional?
```suggestion
    #   "friendly name, phd" <email@code.org>
```

File: bin/cron/deliver_poste_messages [hunk 2]
```diff
@@ -245,20 +245,37 @@ class Deliverer
...
+    #
+    # But we are alerting just in case a sender has attempted to provide multiple
+    # email addresses like this:
+    #   friendly name <email@code.org>, second person <email2@code.org>
+    #
+    # Because we deprecated support for this pattern, believing that we've never
+    # actually used it.
+    if /,(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)/ =~ string
```
Comment (islemaster) [inline bin/cron/deliver_poste_messages]: In the change from your last approach we lost the comment explaining where you got this regex and why it works.  Can we bring that back?

File: bin/cron/deliver_poste_messages [hunk 3]
```diff
@@ -245,20 +245,37 @@ class Deliverer
...
+      Honeybadger.notify(
+        error_message: "Comma outside of surrounding double-quotes detected in address string: #{string}"
+      )
+    end
+
+    if string.include? '<'
+      string =~ /^([^<]*)<([^>]*)>$/
```
Comment (islemaster) [inline bin/cron/deliver_poste_messages]: Certainly a less common case than our existing comma issue, but this doesn't handle the possibility of an escaped or quoted `<` in the display name correctly:
```
"I <3 Code" <iheartcode@example.com>
```
I haven't read the whole standard but based on a preliminary scan of [RFC 2822](https://www.ietf.org/rfc/rfc2822.txt) and [this excellent summary thereof](https://cr.yp.to/immhf/addrlist.html) I don't see anything that would prevent angle brackets in a quoted string `display-name` portion of a `name-addr` token.

If we're having trouble finding code to reuse for this purpose, maybe we should open-source a small gem for it.

File: bin/cron/deliver_poste_messages [hunk 4]
```diff
@@ -245,20 +245,37 @@ class Deliverer
...
+      {name: $1.strip, email: $2}
```
Comment (islemaster) [inline bin/cron/deliver_poste_messages]: In this outcome, `$1.strip` may be a quoted string `"example"` or an unquoted string `example`. In the former case, we'll have stripped outside the quotes but not inside, which might be surprising behavior. `"  example  "`  Do you have a sense of whether it's important to normalize this?

**Review by islemaster — CHANGES_REQUESTED** [review]
Since we're fixing a bug, I'd like to see a few unit tests go in with this change.

**Review by uponthesun — COMMENTED** [review]
Given that we notify HB and continue on when when we find an unquoted comma, what will actually happen with the rest of the code? 

Also, if we can make the fact that we don't support comma-separated recipients explicit to developers writing code to send emails in the future, that'd be great. Perhaps with comments in whatever is the main entry point (poste.rb?), possibly with validation also.

**Comment by uponthesun** [comment]
Also, I just realized `parse_address` is called in many places, are we sure that this change won't affect any of them?

File: bin/cron/deliver_poste_messages (hunk 3 above)
Comment (breville) [reply bin/cron/deliver_poste_messages]: Good call.  Now improved.

**Review by islemaster — APPROVED** [review]
:+1: Better than before and likely to improve over time.  Thanks for doing several rounds on this!

File: bin/cron/deliver_poste_messages (hunk 4 above)
Comment (breville) [reply bin/cron/deliver_poste_messages]: It's a good improvement, but we've lived without it for a few months now, so I'm okay with leaving it unresolved in this PR.

**Comment by breville** [comment]
@uponthesun Great additional questions.

- The path which detects a comma outside of a pair of double-quotes now returns an empty string.
- I've punted on adding additional comments to all of the entry points to keep this change small, but will keep in mind for a future PR.
- `parse_address` is called for six different fields in `deliver_poste_messages`: `reply-to`, `sender`, `from`, `to`, `bcc`, `cc`.  We checked the production database and didn't find any commas outside of double-quotes.  (Except for some `sender` fields from 2014 when it used to store a hash!  That doesn't happen any more.)

**Pull request merged**